code/espurna/e131.cpp

```diff
@@ -0,0 +1,161 @@
+/*
+
+E1.31 MODULE
+
+Copyright (C) 2020 by Adam Honse <calcprogrammer1 at gmail dot com>
+
+*/
+
+#include "e131.h"
+
+#if E131_SUPPORT
+
+#if LIGHT_PROVIDER != LIGHT_PROVIDER_NONE
+
+// -----------------------------------------------------------------------------
+
+#include "light.h"
+#include "web.h"
+#include "ws.h"
+
+#include <ESPAsyncE131.h>
+
+bool         _e131_enabled         = false;
+bool         _e131_multicast       = false;
+unsigned int _e131_universe        = 1;
+bool         _e131_light_0_enabled = false;
+unsigned int _e131_light_0_channel = 1;
+bool         _e131_light_1_enabled = false;
+unsigned int _e131_light_1_channel = 2;
+bool         _e131_light_2_enabled = false;
+unsigned int _e131_light_2_channel = 3;
+bool         _e131_light_3_enabled = false;
+unsigned int _e131_light_3_channel = 4;
+bool         _e131_light_4_enabled = false;
+unsigned int _e131_light_4_channel = 5;
+bool         _e131_initialized     = false;
+
+ESPAsyncE131 e131(2);
+
+// -----------------------------------------------------------------------------
+// E1.31
+// -----------------------------------------------------------------------------
+
+#if WEB_SUPPORT
+
+bool _e131WebSocketOnKeyCheck(const char * key, JsonVariant& value) {
+    return (strncmp(key, "e131", 4) == 0);
+}
+
+void _e131WebSocketOnVisible(JsonObject& root) {
+    root["e131Visible"] = 1;
+}
+
+void _e131WebSocketOnData(JsonObject& root) {
+
+}
+
+void _e131WebSocketOnConnected(JsonObject& root) {
+    root["e131Enabled"]         = getSetting("e131Enabled",       false);
+    root["e131Multicast"]       = getSetting("e131Multicast",     false);
+    root["e131Universe"]        = getSetting("e131Universe",      1);
+    root["e131Light0Enabled"]   = getSetting("e131Light0Enabled", false);
+    root["e131Light0Channel"]   = getSetting("e131Light0Channel", 1);
+    root["e131Light1Enabled"]   = getSetting("e131Light1Enabled", false);
+    root["e131Light1Channel"]   = getSetting("e131Light1Channel", 2);
+    root["e131Light2Enabled"]   = getSetting("e131Light2Enabled", false);
+    root["e131Light2Channel"]   = getSetting("e131Light2Channel", 3);
+    root["e131Light3Enabled"]   = getSetting("e131Light3Enabled", false);
+    root["e131Light3Channel"]   = getSetting("e131Light3Channel", 4);
+    root["e131Light4Enabled"]   = getSetting("e131Light4Enabled", false);
+    root["e131Light4Channel"]   = getSetting("e131Light4Channel", 5);
```

Comment:
*(again about the cfg)*
We also will send less keys to the WebUI.

But, it will be more complicated from the .js script side :/
Another alternative to map string is to use `e131Ch#` and simply loop all channels like Domoticz, Thingspeak map internal IDs to external.

Reply:
Would it be possible to keep the same web UI and pack the settings from it into a string?  I basically just copy pasted the web functionality from other pages as I'm not really sure how it works, nor am I well versed in javascript.

Reply:
Right now this sends key-values as-is, so every key is 'mapped' to the same 'input name=...' elem, this is done at the end of `processData()` func of custom.js

What I meant about `e131Ch#` approach is to do something like:
```cpp
String base("e131Ch");
for (unsigned index = 0; index < lightChannels(); ++index) {
    String key(base + index);
    root[key] = getSetting(key); // without 'default' 2nd param, always returns String
}
```
Treating empty string as disabled and non-empty as e1.31 channel number, while always creating N-channels items. Same thing in the configuration.

Then, we either create html block programmatically like dczRelayTemplate, or always have html blocks in the index.html but hide them by default via `display: none` css style and add additional js code such as:
```js
// processData() near the top...

if (key.startsWith("e131Ch")) {
    $(key).show();
}

// ...rest of processData()
```